src/mutation.cpp

```diff
@@ -463,17 +733,114 @@ void player::activate_mutation( std::string mut )
             newit = i_add(newit);
             add_msg(m_info, "%c - %s", newit.invlet == 0 ? ' ' : newit.invlet, newit.tname().c_str());
         }
-        traits[mut].powered = false;
+        return true;
+    }
+    
+    // Mutation activation was not canceled in legacy mutation effects, so return true.
+    return true;
+}
+
+void Character::deactivate_mutation(const muttype_id &flag)
+{
+    if (!has_mut(flag)) {
+        // We don't have the desired mutation!
+        debugmsg("Tried to deactivate a nonexistent mutation!");
+        return;
+    }
+    if (!my_mutations[flag].is_activatable()) {
+        // It's not an activatable mutation!
+        debugmsg("Tried to deactivate a non-activatable mutation!");
+        return;
+    }
+    if (!my_mutations[flag].is_active()) {
+        // It's already deactivated, abort
         return;
     }
+    if (is_player()) {
+        add_msg(m_neutral, _("You stop using your %s."), my_mutations[flag].get_name().c_str());
+    }
+    
+    // We've decided to deactivate here!
+    // Deactivate the mutation
+    my_mutations[flag].deactivate(*this)
+}
+
+void mutation::deactivate(Character &ch)
+{
+    muttype_id id = mut_type->id;
+    if (deactivation_effects(id, ch)) {
+        // We didn't stop the deactivation
+        charge = 0;
+    }
+}
+
+void mutation::deactivation_effects(const muttype_id id, Character &ch)
+{
+    /* NEW ON-DEACTIVATION MUTATION EFFECTS SHOULD GO HERE.
+     * Cast ch to player if you really need to access player class variables. */
 }
 
-void player::deactivate_mutation(std::string mut)
+void Character::process_mutations()
 {
-    traits[mut].powered = false;
+    // Bitset is <hunger, thirst, fatigue>
+    std::bitset<3> stops;
+    if (hunger >= 2800) {
+        stops.set(1);
+    }
+    if (thirst >= 520) {
+        stops.set(2);
+    }
+    if (fatigue >= 575) {
+        stops.set(3);
+    }
+    std::unordered_map<std::string, int> pro_costs;
+    for (auto &mut : my_mutations) {
+        if (!mut.second.process(pro_costs, stops) && is_player()) {
+            add_msg(m_info, _("You can't keep your %s active anymore!"), mut.get_name().c_str());
+        }
+    }
+    // Pay any reactivation costs
+    hunger += pro_costs["hunger"];
+    thirst += pro_costs["thirst"];
+    fatigue += pro_costs["fatigue"];
```

Comment:
> Once you get a larger number of possible costs I didn't want to have to store a 0 for every single cost for every single mutation (which we already have a ton of, and that number was only going to go up.

Are you concerned about memory consumption? If so, consider that the string itself must be stored (number of characters + 0-char ("hunger"=6) + metadata (length=4) and the map entry is at least 3 pointer (left, right, parent node =3*4)) and the int itself (=4 bytes) gives at least 26 bytes per map entry (entries are allocated on the heap = additional cost for housekeeping by malloc).

However this is totally pointless anyway because whenever you write `costs["thirst"]` (when you use the array operator on the map), the requested entry is created if it not already exists.
```C++
std::map<std::string, int> c;
if(c["hunger"] > 0) { ... }
if(c["thirst"] > 0) { ... }
if(c["fatigue"] > 0) { ... }
// At this point c contains 3 entries: "hunger", "thirst" and "fatigue"
```
You end up with completely filled maps anyway.

Or is it about initialization? By using the in-class initialization, you only have to do it once anyway.

However when you add another cost, look at all the places that you have to change anyway (even with the current usage of the map): basically everywhere where the map is used.

----

Storing it in a map seems to yield no advantage in this case (advantage would be for example if the number of costs can be quite huge, or if you're going to iterate over them/search for an entry with a dynamically created key). You're just reinventing a `struct`. But how about a middle way: use static const strings instead of string literals as keys:
```C++
static const std::string HUNGER_COST_ID("hunger");

costs[HUNGER_COST_ID] += ...;
```
That gives you the advantages of the map and the safety of the compiler checking for misspelled keys.

----

I have another, nicer but also more complex solution - the beauty of C++ (-:
```C++
struct cost_type {
    std::string id;
    std::string name; // display name
    virtual void consume(Charachter&, int) const = 0;
    virtual bool can_consume(const Charachter &c) const = 0;
};

struct hunger_cost : public cost_type {
    void consume(Charachter &c, int cost) const {
        c.hunger += cost;
    }
    bool can_consume(const Charachter &c) const {
        return c.hunger <= 500;
    }
};

static const hunger_cost hunger_cost( "hunger", _("hunger") );
// similar for thirst etc. Might also work with simple
// std::function<...> from lambdas instead of virtual functions
static const std::vector<const cost_type*> cost_types = { &hunger_cost, &thirst_cost, ... };

std::map<const cost_type*, int> costs;
void load_cost(JsonObject jo, std::string ide) {
    cost_type *ct = find_cost_type(id);
    costs[id] = joget_int("cost");
}
void consume() {
    for(auto &e : costs) {
        e.first->consume(g->u, e.second);
    }
}
```
While loading from json, use the key of the cost map to search for a cost_type object in cost_types that has the same id and use the pointer to that object as key in the map. Than you can really iterate over the cost map all the time instead of extracting specific entries.

This keeps everything nicely together: handling "hunger cost" is completely separated from fatigue cost etc. Everything is type safe and checked for consistency.